File: 8. Graph/Adjacency Matrix/Un-Directed Graph/unDirectedGraph.cpp
#include<iostream>
using namespace std;

struct Node {
    int data;
    struct Node* next;
};

int** createGraph(int vertex) {
    int** graph = (int**) malloc(vertex * sizeof(int*));

    for (int i = 0; i < vertex; i++) {
        graph[i] = (int*)malloc(vertex * sizeof(int));
    }
    
    return graph;
}

struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*) malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

void addEdge(int** graph, int src, int data) {
    graph[src][data] = 1;
    return;
}

void bfsTraversal(int ** graph, int startVertex, int *visited, int vertex) {
    struct Node* queue = createNode(startVertex);
    struct Node* lastQueueNode = queue;

    if (visited[startVertex]) {
        return;
    }

    while (queue != NULL) {
        int current = queue->data;
        cout << current << " ";
        visited[startVertex] = 1;
        

        for (int neighbour = 0; neighbour < vertex; neighbour++) {
            int value = graph[current][neighbour];

            if (value && !visited[neighbour]) {
                struct Node *node = createNode(neighbour);
                visited[neighbour] = 1;
                lastQueueNode->next = node;
                lastQueueNode = lastQueueNode->next;
            }
            
        }

        queue = queue->next;
    }

    return;
}

void dfsTraversal(int ** graph, int startVertex, int *visited, int vertex) {
    if (visited[startVertex]) {
        return;
    }
    
    cout << startVertex << " ";
    visited[startVertex] = 1;

    for (int neighbour = 0; neighbour < vertex; neighbour++) {
        int value = graph[startVertex][neighbour];

        if (value && !visited[neighbour]) {
            dfsTraversal(graph, neighbour, visited, vertex);
        }
    }

    return;
}

void findAllPaths(int ** graph, int *path, int startVertex, int endVertex, int index, int vertex, int* visited) {
    visited[startVertex] = 1;
    path[index++] = startVertex;

    if (startVertex == endVertex) {
        for (int i = 0; i < index; i++) {
            if (i == index - 1) {
                cout << path[i];
            } else {
                cout << path[i] << " -> ";
            }
            
        }
        cout << "\n";
        return;
    }

    for (int neighbour = 0; neighbour < vertex; neighbour++) {
        int value = graph[startVertex][neighbour];

        if (value) {
            if (!visited[neighbour]) {
                visited[neighbour] = 1;
                findAllPaths(graph, path, neighbour, endVertex, index, vertex, visited);
                visited[neighbour] = 0;
            }
            
        }
    }

    return;
}

int isCycleExistsUtils(int ** graph, int startVertex, int* visited, int parent, int vertex) {
    visited[startVertex] = 1;

    for(int neighbour = 0; neighbour < vertex; neighbour++) {
        int value = graph[startVertex][neighbour];

        if (value && !visited[neighbour]) {
            if (isCycleExistsUtils(graph, neighbour, visited, startVertex, vertex)) {
                return 1;
            }
        } else if (value && neighbour != parent) {
            return 1;
        }
    }

    return 0;
}

int isCycleExists(int** graph, int vertex) {
    int cycleExists = 0;
    int *visited = (int *)malloc(vertex * sizeof(int));

    cout << "\n";

    for (int i = 0; i < vertex; i++) {
        if (!visited[i]) {
            if (isCycleExistsUtils(graph, i, visited, -1, vertex)) {
                cycleExists = 1;
                break;
            }
        }
        
    }

    free(visited);
    return cycleExists;
}

int main() {

    cout << "Hello! Here You Can Perform Following Directed Graph Operation!\n";
    
    int vertex;
    cout << "\nEnter the Number of Vertices(Node): ";
    cin >> vertex;
    
    int** graph = createGraph(vertex);

    int edges;

    cout << "\nEnter the Number of Edges: ";
    cin >> edges;

    for (int i = 0; i < edges; i++) {
        int src, dest;

        cout << "\nEnter the Source of Edge " << i + 1 << ": ";
        cin >> src;

        cout << "Enter the Destination of Edge " << i + 1 << ": ";
        cin >> dest;

        if (src > vertex - 1) {
            cout << "\n~~ " << src << " Source Vertex Not Present ~~\n\n";
            i--;
            continue;
        }
        
        if (dest > vertex - 1) {
            cout << "\n~~ " << dest << " Target Vertex Not Present ~~\n\n";
            i--;
            continue;
        }
        
        if (src == dest) {
            cout << "\n~~ Source & Target Can't Be Same ~~\n\n";
            i--;
            continue;
        }

        addEdge(graph, src, dest);
        addEdge(graph, dest, src);
    }

    int choice;

    while (1) {
        cout << "\n1. Breadth First Search\n2. Depth First Search\n3. Find All Path's\n4. Is Cycle Exists\n5. Exit\nEnter Your Choice:- ";
        cin >> choice;

        switch (choice) {
        case 1: {
            int *visited = (int *)malloc(vertex * sizeof(int));

            cout << "\n";

            for (int i = 0; i < vertex; i++) {
                bfsTraversal(graph, i, visited, vertex);
            }

            cout << "\n";

            free(visited);

            break;
        }

        case 2: {
            int *visited = (int *)malloc(vertex * sizeof(int));

            cout << "\n";

            for (int i = 0; i < vertex; i++) {
                dfsTraversal(graph, i, visited, vertex);
            }

            cout << "\n";

            free(visited);

            break;
        }

        case 3: {
            int startVertex, endVertex, index = 0;
            int *path = (int *)malloc(2 * vertex * sizeof(int));
            int *visited = (int *)malloc(vertex * sizeof(int));

            cout << "\nEnter Start Vertex: ";
            cin >> startVertex;
            
            cout << "Enter End Vertex: ";
            cin >> endVertex;
            
            cout << "\nPaths: \n";
            findAllPaths(graph, path, startVertex, endVertex, index, vertex, visited);

            free(path);
            free(visited);
            
            break;
        }

        case 4: {
            if (isCycleExists(graph, vertex)) {
                cout << "Cycle Exists!\n";
            } else {
                cout << "Cycle Does Not Exists!\n";
            }

            break;
        }

        case 5:
            cout << "Thanks For Using! Have a Great Day!";
            free(graph);
            return 0;
        default:
            cout << "Please Enter a Vaild Input!\n";
            break;
        }
    }

    return 0;
}

File: 8. Graph/Adjacency Matrix/Un-Directed Graph/unDirectedGraph.js
const prompt = require('prompt-sync')({ sigint: true });

class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
};

function addEdge(graph, src, data) {
    graph[src][data] = 1;
    return;
}

function bfsTraversal(graph, startVertex, visited, vertex, printArr) {
    let queue = new Node(startVertex);
    let lastQueueNode = queue;

    if (visited[startVertex]) {
        return;
    }

    while (queue != null) {
        let current = queue.data;

        printArr.push(current)
        visited[startVertex] = true;

        for (let neighbour = 0; neighbour < vertex; neighbour++) {
            let value = graph[current][neighbour];

            if (value == 1 && !visited[neighbour]) {
                let node = new Node(neighbour);
                visited[neighbour] = true;
                lastQueueNode.next = node;
                lastQueueNode = lastQueueNode.next;
            }

        }

        queue = queue.next;
    }

    return;
}

function dfsTraversal(graph, startVertex, visited, vertex, printArr) {
    if (visited[startVertex]) {
        return;
    }

    printArr.push(startVertex);
    visited[startVertex] = true;

    for (let neighbour = 0; neighbour < vertex; neighbour++) {
        let value = graph[startVertex][neighbour];

        if (value == 1 && !visited[neighbour]) {
            dfsTraversal(graph, neighbour, visited, vertex, printArr);
        }
    }

    return;
}

function findAllPaths(graph, path, startVertex, endVertex, vertex, visited) {
    visited[startVertex] = true;
    path.push(startVertex);

    if (startVertex == endVertex) {
        console.log(path.join(" -> "));
        path.pop();
        return;
    }

    for (let neighbour = 0; neighbour < vertex; neighbour++) {
        let value = graph[startVertex][neighbour];

        if (value == 1) {
            if (!visited[neighbour]) {
                visited[neighbour] = true;
                findAllPaths(graph, path, neighbour, endVertex, vertex, visited);
                visited[neighbour] = false;
            }
        }
    }

    path.pop();

    return;
}

function isCycleExistsUtils(graph, startVertex, visited, parent, vertex) {
    visited[startVertex] = true;

    for (let neighbour = 0; neighbour < vertex; neighbour++) {
        let value = graph[startVertex][neighbour];

        if (value == 1 && !visited[neighbour]) {
            if (isCycleExistsUtils(graph, neighbour, visited, startVertex, vertex)) {
                return true;
            }
        } else if (value == 1 && neighbour != parent) {
            return true;
        }
    }

    return false;
}

function isCycleExists(graph, vertex) {
    let cycleExists = false;
    let visited = Array.from({ length: vertex }, () => false);

    for (let i = 0; i < vertex; i++) {
        if (!visited[i]) {
            if (isCycleExistsUtils(graph, i, visited, -1, vertex)) {
                cycleExists = true;
                break;
            }
        }
    }

    return cycleExists;
}

(() => {
    console.log("Hello! Here You Can Perform Following Directed Graph Operation!\n");
    let vertex = prompt("Enter the Number of Vertices(Node): ");

    let graph = Array.from({ length: vertex }, () => []);

    let edges = prompt("Enter the Number of Edges: ");

    for (let i = 0; i < edges; i++) {
        console.log();
        let src = Number.parseInt(prompt(`Enter the Source of Edge ${i + 1}: `));
        let dest = Number.parseInt(prompt(`Enter the Source of Target ${i + 1}: `));

        if (src > graph.vertex - 1) {
            console.log(`~~ ${src} Source Vertex Not Present ~~\n`);
            i -= 1;
            continue;
        }

        if (dest > graph.vertex - 1) {
            console.log(`~~ ${dest} Target Vertex Not Present ~~\n`);
            i -= 1;
            continue;
        }

        if (src == dest) {
            console.log(`~~ Source & Target Can't Be Same ~~\n`);
            i -= 1;
            continue;
        }

        addEdge(graph, src, dest);
        addEdge(graph, dest, src);
    }

    while (true) {
        console.log("\n1. Breadth First Search\n2. Depth First Search\n3. Find All Path's\n4. Is Cycle Exists\n5. Exit");
        choice = Number.parseInt(prompt("Enter Your Choice:- "));

        switch (choice) {
            case 1:
                var visited = Array.from({ length: vertex }, () => false);
                var printArr = [];

                console.log();

                for (let j = 0; j < vertex; j++) {
                    if (!visited[j]) {
                        bfsTraversal(graph, j, visited, vertex, printArr);
                    }
                }

                console.log(printArr.join(" -> "));

                continue;

            case 2:
                var visited = Array.from({ length: vertex }, () => false);
                var printArr = [];

                console.log();

                for (let j = 0; j < vertex; j++) {
                    if (!visited[j]) {
                        dfsTraversal(graph, j, visited, vertex, printArr);
                    }
                }

                console.log(printArr.join(" -> "));

                continue;

            case 3:
                let path = [];
                var visited = Array.from({ length: vertex }, () => false);

                console.log();

                let startVertex = Number.parseInt(prompt("Enter Start Vertex: "));
                let endVertex = Number.parseInt(prompt("Enter End Vertex: "));

                console.log("\nPaths: ");

                findAllPaths(graph, path, startVertex, endVertex, vertex, visited);

                continue;

            case 4: 
                if (isCycleExists(graph, vertex)) {
                    console.log("\nCycle Exists");
                } else {
                    console.log("\nCycle Does Not Exists");
                }

                break;

            case 5:
                console.log("Thanks For Using! Have a Great Day!");
                return;
            default:
                console.log("Please Enter a Vaild Input!\n");
                continue;
        }
    }
    return;
})();